### PR TITLE
storage: add a grpc remote storage

### PR DESCRIPTION
**Pull request by huachaohuang**
This PR only implements the client part of the storage and leaves the server part to future PRs.

This storage uses gRPC for communications. Since the storage APIs are very simple, maybe we can find a more efficient way for RPC in the future. As long as the user interface remains the same, we will be fine.

The API design mostly follows the [Google API design guide](https://cloud.google.com/apis/design) and the [googleapis](https://github.com/googleapis/googleapis/blob/master/google/storage/v2/storage.proto), which is very nice and worth reading. However, most part of the current implementation is suboptimal in order to keep things simple for now.

### CODE REVIEW

**Review by tisonkun — APPROVED** [review]
The code side looks good to me. I'd like to ask what components will implement as `storage_server`? From the design I can see a "Storage Service" but is it another storage unit?

Also I like `method!` macro and be curious that it isn't implemented in tonic or other libs.

**Comment by huachaohuang** [comment]
@tisonkun The idea is to let different components provide a general type of [tower service](https://docs.rs/tower-service/0.1.0/tower_service/trait.Service.html), which has a good Rust ecosystem. For example, the grpc storage will provide a grpc storage service. And most units will be a general-purpose HTTP server that can serve one or more such kind of services. For example, a storage unit will be an HTTP server that serves a storage service.

So the `Unit/UnitBuilder` trait in microunit will be the bridge between different units and services. Then we may need to decide where these unit implementations should go. For example, we can keep the components totally unaware of microunit by placing unit implementations in a middle layer.

As for the `method!` macro, I think it may only work for now. As the implementation becomes more complicated, different methods may handle things differently. So I think `method!` will be removed sooner or later :)

**Comment by huachaohuang** [comment]
I will try to implement the grpc storage service and unit next to see if these things work together. Then I will update the design documents.

File: src/storage/src/grpc/bucket.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        let input = ListObjectsRequest {
+            bucket: self.bucket.clone(),
+        };
+        match self.client.list_objects(input).await {
+            Ok(output) => {
+                let object_names = output
+                    .objects
+                    .into_iter()
+                    .map(Ok)
+                    .collect::<Vec<Result<String>>>();
```
Comment (zojw) [inline src/storage/src/grpc/bucket.rs:52]: we also can remove this line.

File: src/storage/src/grpc/client.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+impl Client {
+    method!(list_buckets, ListBucketsRequest, ListBucketsResponse);
+
+    method!(create_bucket, CreateBucketRequest, CreateBucketResponse);
+
+    method!(delete_bucket, DeleteBucketRequest, DeleteBucketResponse);
+
+    method!(list_objects, ListObjectsRequest, ListObjectsResponse);
+
+    method!(upload_object, UploadObjectRequest, UploadObjectResponse);
```
Comment (zojw) [inline src/storage/src/grpc/client.rs:46]: for upload/read object maybe we can using [grpc stream](https://github.com/hyperium/tonic/blob/master/examples/routeguide-tutorial.md#client-side-streaming-rpc) , but it seems be complex and can do it later~

**Review by zojw — APPROVED** [review]
rest LGTM

File: src/storage/src/grpc/client.rs (hunk 2 above)
Comment (huachaohuang) [reply src/storage/src/grpc/client.rs:46]: Yes, I want to do that too. But I just try to make things work to see how they fit together for now :)

File: src/storage/src/grpc/storage.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    async fn list_buckets(&self) -> ResultStream<String> {
+        let input = ListBucketsRequest {};
+        match self.client.list_buckets(input).await {
+            Ok(output) => {
+                let bucket_names = output
+                    .buckets
+                    .into_iter()
+                    .map(Ok)
+                    .collect::<Vec<Result<String>>>();
```
[the remark this answers is not in the document]
Comment (huachaohuang) [reply src/storage/src/grpc/storage.rs:50]: Oh, really!? Let me give it a try.

File: src/storage/src/grpc/storage.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            }
+            Err(err) => Box::new(stream::once(future::err(err.into()))),
+        }
+    }
+
+    async fn create_bucket(&self, name: &str) -> Result<Box<dyn Bucket>> {
+        let input = CreateBucketRequest {
+            bucket: name.to_owned(),
+        };
+        let _ = self.client.create_bucket(input).await?;
```
[the remark this answers is not in the document]
Comment (huachaohuang) [reply src/storage/src/grpc/storage.rs:61]: I just assume that it is required to compile. Let me see if I can remove it.

File: src/storage/src/grpc/object.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    async fn read_at(&self, buf: &mut [u8], offset: usize) -> Result<usize> {
+        let input = ReadObjectRequest {
+            bucket: self.bucket.clone(),
+            object: self.object.clone(),
+            offset: offset as i64,
+            length: buf.len() as i64,
+        };
+        let output = self.client.read_object(input).await?;
+        let content = output.content;
+        // TODO: refine the interface to avoid copy here.
```
Comment (tisonkun) [inline src/storage/src/grpc/object.rs:57]: If the parameter is `&mut [u8]`, we must copy the value from output to the empty buffer.

@huachaohuang do you want to pass a pointer of reference, or is there zero-copy technique to mapping one buffer to another?

Comment (huachaohuang) [reply src/storage/src/grpc/object.rs:57]: Well, we can change the interface to return a buffer. But in some cases, callers may have a buffer at handed and just want us to copy the data into their buffer, then returning a buffer to them is not friendly. So maybe we need to find a way that satisfies different use cases. Or we simply provide different interfaces. I don't know yet.

Comment (zojw) [reply src/storage/src/grpc/object.rs:57]: maybe we can 
```
async fn read_at<'a>(&self, offset: usize, len: usize, opt_buf: &'a mut [u8]) -> (Result<usize>, Cow<'a, &[u8]>)
```
and fill and return `opt_buf` if user provide it, and return rpc buffer if not.

but in this way, we cannot metaphor `len` information from buf.len() any more...- -

Comment (tisonkun) [reply src/storage/src/grpc/object.rs:57]: @huachaohuang OK. You can create an issue post-v0.2. In this case I'd prefer:

1. Provide two interfaces for each use case.
2. Provide one interface returning buffer and let users decide what to do.
3. Provide the current interface with a known and bearable overhead, at least for early releases.

@zojw I don't think an interface in that signature a good idea - it's too complex to understand and anti-intuitive.

**Pull request merged**